File: BTPreIn.java
// Time Complexity : O(n*2)
// Space Complexity : O(n*2) including recursive stack, else O(n) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


//We find root node from preorder first element
//loop through inorder to find the root, so all the elements left of root form left subtree and right of root form right subtree
//we build 4 arrays, 2 left and right for both inorder and preorder and recursively go till the leaf nodes to build tree
//return root in the last


class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if(preorder.length==0) {
            return null;
        }
        
        TreeNode root = new TreeNode(preorder[0]);
        int index =-1;
        for(int i=0;i<inorder.length;i++){
            if(inorder[i]==root.val){
                index=i;
                
            }
        }
        
         int[] leftPre=Arrays.copyOfRange(preorder,1,index+1);
        int[] leftIn=Arrays.copyOfRange(inorder,0,index);
        
         int[] rightPre=Arrays.copyOfRange(preorder,index+1,preorder.length);
        int[] rightIn=Arrays.copyOfRange(inorder,index+1,inorder.length);
       
       
        
        root.left=buildTree(leftPre,leftIn);
        root.right=buildTree(rightPre,rightIn);
        
        return root;
        
    }
}

File: bt_pre_in.py
#Time Complexity: O(N).
#Auxiliary Space: O(N)
#Did this code successfully run on Leetcode :Yes

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def __init__(self):
        self.pre_idx =0
        self.inorder_map = {}
        
    
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        
        def array_tree(left,right):
            
            if left>=right:
                return
        
            root_value = preorder[self.pre_idx]
            root_index = self.inorder_map[root_value]
            root = TreeNode(root_value)
            self.pre_idx+=1
            root.left = array_tree(left,root_index)
            root.right = array_tree(root_index+1,right)
            return root
        
        
        for index, value in enumerate(inorder):
            self.inorder_map[value] = index
            
        
        return array_tree(0,len(preorder))